automates/program_analysis/PyAST2CAST/py_ast_to_cast.py

```diff
@@ -0,0 +1,1321 @@
+                    source_refs=ref
+                )]
+        else:
+            loop_cond = BinaryOp(
+                BinaryOperator.LT,
+                Name(var_name,source_refs=ref),
+                Call(Name("len",source_refs=ref), [iterable], source_refs=ref),
+                source_refs=ref
+            )
+            if isinstance(node.target,ast.Tuple):
+                loop_assign = [Assignment(
+                    Tuple([Var(type="integer",val=Name(node.val.name+"_",source_refs=ref),source_refs=ref) for node in target.values],source_refs=ref),
+                    Subscript(Name(node.iter.id,source_refs=ref), Name(var_name,source_refs=ref),source_refs=ref),
+                    source_refs=ref
+                ),
+                Assignment(
+                    target,
+                    Tuple([Var(type="integer",val=Name(node.val.name+"_",source_refs=ref),source_refs=ref) for node in target.values],source_refs=ref),
+                    source_refs=ref
+                )]
+            else:
+                loop_assign = [Assignment(
+                    Var(Name(node.target.id, source_refs=ref), "integer", source_refs=ref),
+                    Subscript(Name(node.iter.id, source_refs=ref), Name(var_name, source_refs=ref), source_refs=ref),
+                    source_refs=ref
+                )]
+        loop_increment = [Assignment(
+            Var(Name(var_name, source_refs=ref), "integer", source_refs=ref),
+            BinaryOp(BinaryOperator.ADD, Name(var_name, source_refs=ref), Number(1, source_refs=ref), source_refs=ref),
+            source_refs=ref
+        )]
+
+        return [count_var, Loop(expr=loop_cond, body=loop_assign + body + loop_increment, source_refs=ref)]
+
+    def visit_FunctionDef(self, node: ast.FunctionDef):
+        """Visits a PyAST FunctionDef node. Which is used for a Python
+        function definition.
+
+        Args:
+            node (ast.FunctionDef): A PyAST function definition node
+
+        Returns:
+            FunctionDef: A CAST Function Definition node
+        """
+
+        body = []
+        args = []
+        # TODO: Correct typing instead of just 'integer'
+        if len(node.args.args) > 0:
+            args = [Var(Name(arg.arg,source_refs=[SourceRef(self.filenames[-1], arg.col_offset, arg.end_col_offset, arg.lineno, arg.end_lineno)]), 
+                    "integer", 
+                    source_refs=[SourceRef(self.filenames[-1], arg.col_offset, arg.end_col_offset, arg.lineno, arg.end_lineno)]) 
+                    for arg in node.args.args]
+        if len(node.body) > 0:
+            # To account for nested loops we check to see if the CAST node is in a list and
+            # extend accordingly
+            for piece in node.body:
+                to_add = self.visit(piece)
+                if isinstance(to_add,Module):
+                    body.extend([to_add])
+                else:
+                    body.extend(to_add)
```

Comment:
Well, every visitor besides the Module visitor that returns CAST nodes returns them in a list of CAST nodes, even if the list is a one element list. No other visitor that I can recall does that. The reason Module is like this is because it can both be the starting point of a program or be imported by another program. That's why this check is here. I like that extra safety check so I'm going to incorporate it. Thank you for the suggestion!